File: app/src/main/java/com/returnnotfound/stain/widget/image/ImageLoader.java
package com.returnnotfound.stain.widget.image;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.widget.ImageView;

import com.bumptech.glide.RequestBuilder;
import com.returnnotfound.stain.App;
import com.returnnotfound.stain.R;
import com.returnnotfound.stain.utils.ContextUtils;

public interface ImageLoader {
  String FILE_EXTENSION = ".JPG";
  int DEFAULT_PLACE_HOLDER = R.drawable.image_placeholder;
  String CACHE_PATH = App.Companion.getInstance().getCacheDir().getAbsolutePath();

  RequestBuilder<Drawable> getRequestBuilderDefault(ImageView imageView);

  // Load with String
  void loadImage(ImageView imageView, String image);

  void loadImage(ImageView imageView, String image, String imageThumbnail);

  void loadImage(ImageView imageView, String image, int placeHolderId, int errorId);

  void loadImage(ImageView imageView, String image, String thumbnail, int placeHolderId, int errorId);

  // Load with intRes
  void loadImage(ImageView imageView, int imageResId);

  void loadImage(ImageView imageView, int imageResId, int placeHolderId, int errorId);

  // Load with Uri
  void loadImage(ImageView imageView, Uri uri);

  void loadImage(ImageView imageView, Uri uri, int placeHolderId, int errorId);

  RequestBuilder<Drawable> getRequestBuilder(ImageView imageView, ImageLoaderGlide.LOAD_TYPE loadType, String image, int imageResId, Uri uri, String thumbnail, int placeHolderId, int errorId);

  void pauseLoad(Context context);

  void resumeLoad(Context context);

  default boolean isValid(ImageView imageView) {
    return imageView != null && (ContextUtils.isValidContext(imageView.getContext()));
  }
}

File: app/src/main/java/com/returnnotfound/stain/widget/image/ImageLoader.kt
package com.returnnotfound.stain.widget.image

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.widget.ImageView
import com.bumptech.glide.RequestBuilder
import com.returnnotfound.stain.App
import com.returnnotfound.stain.R
import com.returnnotfound.stain.utils.ContextUtils

interface ImageLoader {
  fun getRequestBuilderDefault(imageView: ImageView?): RequestBuilder<Drawable?>?

  fun loadImage(
    imageView: ImageView? = null,
    url: String? = null,
    resId: Int = DEFAULT_PLACE_HOLDER,
    uri: Uri? = null,
    thumbnail: String? = null,
    placeHolderId: Int = DEFAULT_PLACE_HOLDER,
    errorId: Int = DEFAULT_PLACE_HOLDER
  )

  fun getRequestBuilder(
    imageView: ImageView? = null,
    url: String? = null,
    resId: Int = DEFAULT_PLACE_HOLDER,
    uri: Uri? = null,
    thumbnail: String? = null,
    placeHolderId: Int = DEFAULT_PLACE_HOLDER,
    errorId: Int = DEFAULT_PLACE_HOLDER
  ): RequestBuilder<Drawable?>?

  fun pauseLoad(context: Context?)

  fun resumeLoad(context: Context?)

  fun isValid(imageView: ImageView?) =
    imageView != null && ContextUtils.isValidContext(imageView.context)

  companion object {
    const val FILE_EXTENSION = ".JPG"
    const val DEFAULT_PLACE_HOLDER = R.drawable.image_placeholder
    val CACHE_PATH: String = App.instance!!.cacheDir.absolutePath
  }
}